include/zephyr/drivers/gnss.h

```diff
@@ -0,0 +1,429 @@
+/*
+ * Copyright (c) 2023 Trackunit Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file gnss.h
+ * @brief Public GNSS API.
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_GNSS_H_
+#define ZEPHYR_INCLUDE_DRIVERS_GNSS_H_
+
+/**
+ * @brief GNSS Interface
+ * @defgroup gnss_interface GNSS Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <zephyr/device.h>
+#include <zephyr/data/navigation.h>
+#include <errno.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** GNSS PPS modes */
+enum gnss_pps_mode {
+	/** PPS output disabled */
+	GNSS_PPS_MODE_DISABLED = 0,
+	/** PPS output always enabled */
+	GNSS_PPS_MODE_ENABLED = 1,
+	/** PPS output enabled from first lock */
+	GNSS_PPS_MODE_ENABLED_AFTER_LOCK = 2,
+	/** PPS output enabled while locked */
+	GNSS_PPS_MODE_ENABLED_WHILE_LOCKED = 3
+};
+
+/** API for setting fix rate */
+typedef int (*gnss_set_fix_rate_t)(const struct device *dev, uint32_t fix_interval);
+
+/** API for getting fix rate */
+typedef int (*gnss_get_fix_rate_t)(const struct device *dev, uint32_t *fix_interval);
+
+/**
+ * @brief GNSS periodic tracking configuration
+ *
+ * @note Setting either active_time or inactive_time to 0 will disable periodic
+ * function.
+ */
+struct gnss_periodic_config {
+	/** The time the GNSS will spend in the active state in ms */
+	uint32_t active_time;
+	/** The time the GNSS will spend in the inactive state in ms */
+	uint32_t inactive_time;
+};
+
+/** API for setting periodic tracking configuration */
+typedef int (*gnss_set_periodic_config_t)(const struct device *dev,
+					  const struct gnss_periodic_config *periodic_config);
+
+/** API for setting periodic tracking configuration */
+typedef int (*gnss_get_periodic_config_t)(const struct device *dev,
+					  struct gnss_periodic_config *periodic_config);
+
+/** GNSS navigation modes */
+enum gnss_navigation_mode {
+	/** Dynamics have no impact on tracking */
+	GNSS_NAVIGATION_MODE_ZERO_DYNAMICS = 0,
+	/** Low dynamics have higher impact on tracking */
+	GNSS_NAVIGATION_MODE_LOW_DYNAMICS = 1,
+	/** Low and high dynamics have equal impact on tracking */
+	GNSS_NAVIGATION_MODE_BALANCED_DYNAMICS = 2,
+	/** High dynamics have higher impact on tracking */
+	GNSS_NAVIGATION_MODE_HIGH_DYNAMICS = 3
+};
```

Comment:
how standard is this? wondering if this should not be left up to a vendor specific something... though it'd be cool to map to known values like these ones

Reply:
These are purely an attempt of abstraction, not standard to any vendor specific navigation mode. The plan is to allow users to map vendor specific modes to these modes using Kconfig, I see no other way of abstracting these modes as they are just to varied

Reply:
For the LC76G, I mapped them as follows:
```
GNSS_NAVIGATION_MODE_ZERO_DYNAMICS -> STATIONARY
GNSS_NAVIGATION_MODE_LOW_DYNAMICS -> FITNESS
GNSS_NAVIGATION_MODE_BALANCED_DYNAMICS -> NORMAL
GNSS_NAVIGATION_MODE_HIGH_DYNAMICS -> DRONE
```

For the NEO-M8, it could be
```
GNSS_NAVIGATION_MODE_ZERO_DYNAMICS -> STATIONARY
GNSS_NAVIGATION_MODE_LOW_DYNAMICS -> AUTOMOTIVE
GNSS_NAVIGATION_MODE_BALANCED_DYNAMICS -> PORTABLE
GNSS_NAVIGATION_MODE_HIGH_DYNAMICS -> AIRBORNE_2G
```

At least "stationary" and "normal" seem to be quite agreed upon, the dynamic ones are wildly varied